map.go

```diff
@@ -596,11 +589,144 @@ func (m *Map[K, V]) resize(newCapacity uintptr) {
 }
 
 func (m *Map[K, V]) rehashInPlace() {
-	// TODO(peter): rehash in place if there are a sufficient number of
-	// tombstones to reclaim. See drop_deletes_without_resize() in the abseil
-	// implementation:
-	// https://github.com/abseil/abseil-cpp/blob/master/absl/container/internal/raw_hash_set.h#L311
-	panic("TODO(peter): unimplemented")
+	if debug {
+		fmt.Printf("rehash: %d/%d\n", m.used, m.capacity)
+		for i := uintptr(0); i < m.capacity; i++ {
+			switch *m.ctrls.At(i) {
+			case ctrlEmpty:
+				fmt.Printf("  %d: empty\n", i)
+			case ctrlDeleted:
+				fmt.Printf("  %d: deleted\n", i)
+			case ctrlSentinel:
+				fmt.Printf("  %d: sentinel\n", i)
+			default:
+				fmt.Printf("  %d: %v\n", i, m.slots.At(i).key)
+			}
+		}
+	}
+
+	// We want to drop all of the deletes in place. We first walk over the
+	// control bytes and mark every DELETED slot as EMPTY and every FULL slot
+	// as DELETED. Marking the DELETED slots as EMPTY has effectively dropped
+	// the tombstones, but we fouled up the probe invariant. Marking the FULL
+	// slots as DELETED gives us a marker to locate the previously FULL slots.
+
+	// Mark all DELETED slots as EMPTY and all FULL slots as DELETED.
+	for i := uintptr(0); i < m.capacity; i += groupSize {
+		m.ctrls.At(i).convertDeletedToEmptyAndFullToDeleted()
+	}
+
+	// Fixup the cloned control bytes and the sentinel.
+	for i, n := uintptr(0), uintptr(groupSize-1); i < n; i++ {
+		*m.ctrls.At(((i - (groupSize - 1)) & m.capacity) + (groupSize - 1)) = *m.ctrls.At(i)
+	}
+	*m.ctrls.At(m.capacity) = ctrlSentinel
+
+	// Now we walk over all of the DELETED slots (a.k.a. the previously FULL
+	// slots). For each slot we find the first probe group we can place the
+	// element in which reestablishes the probe invariant. Note that as this
+	// loop proceeds we have the invariant that there are no DELETED slots in
+	// the range [0, i). We may move the element at i to the range [0, i) if
+	// that is where the first group with an empty slot in its probe chain
+	// resides, but we never set a slot in [0, i) to DELETED.
+	for i := uintptr(0); i < m.capacity; i++ {
+		if *m.ctrls.At(i) != ctrlDeleted {
+			continue
+		}
+
+		s := m.slots.At(i)
+		h := m.hash(noescape(unsafe.Pointer(&s.key)), m.seed)
+		seq := makeProbeSeq(h1(h), m.capacity)
+		desired := seq
+
+		probeIndex := func(pos uintptr) uintptr {
+			return ((pos - desired.offset) & m.capacity) / groupSize
+		}
+
+		for ; ; seq = seq.next() {
```

Comment:
[nit] I think we can just loop until we find a non-zero `match` and do everything else after the loop.

Reply:
Good idea. Done.